examples/demo/public/lib/loc.js

```diff
@@ -96,33 +104,54 @@ export function Router(props) {
 		});
 
 		return h(Committer, {}, h(RouteContext.Provider, { value: m }, p || d));
-	}, [url]);
-
-	this.componentDidCatch = err => {
-		if (err && err.then) pending.current = err;
-	};
+	}, [url, path, query]);
 
 	useLayoutEffect(() => {
 		let p = pending.current;
+		let isCurrentCycle = true;
 
 		const commit = () => {
-			if (cur.current.url !== url || pending.current !== p) return;
-			prev.current = prevChildren.current = pending.current = null;
-			if (props.onLoadEnd) props.onLoadEnd(url);
-			update(0);
+			if (cur.current.url !== url || pending.current !== p || !isCurrentCycle) {
+				let p, d, m;
+				[].concat(props.children || []).some(vnode => {
+					const matches = exec(path, vnode.props.path, (m = { path, query }));
+					if (matches) return (p = cloneElement(vnode, m));
+					if (vnode.props.default) d = cloneElement(vnode, m);
+				});
+
+				cur.current.url = url;
+				prevChildren.current = null;
+				curChildren.current = h(Committer, {}, h(RouteContext.Provider, { value: m }, p || d));
+				update(0);
+				return;
+			} else {
+				prev.current = prevChildren.current = pending.current = null;
+				if (props.onLoadEnd) props.onLoadEnd(url);
+				update(0);
+				if (wasPush) scrollTo(0, 0);
+			}
 		};
 
 		if (p) {
 			if (props.onLoadStart) props.onLoadStart(url);
 			p.then(commit);
-		} else commit();
-	}, [url]);
+		} else {
+			commit();
+		}
 
-	// Hi! Wondering what this horrid line is for? That's totally reasonable, it is gross.
-	// It prevents the old route from being remounted because it got shifted in the children Array.
-	if (reverse && this.__v && this.__v.__k) this.__v.__k.reverse();
+		return () => {
+			if (cur.current.url !== url || pending.current !== p) {
```

Comment:
Rather than synchronously checking whether the url has changed we need to do this within the effect, the url and mutative ref could be equal again in a scenario where we `prerender --> lazy-load --> switch` which would mean that we push the oldChildren back up